juspay/hyperswitch
# feat(payment_methods): add support to migrate existing customer PMs from processor to hyperswitch

## Type of Change



## Description

Merchants who has existing customer payment methods in payment processor can upload a csv file to migrate their customers and their payment methods using this `/migrate-bulk` API

### Additional Changes





## Motivation and Context

To provide seamless experience to customers when merchant moves from payment processor to hyperswitch

## How did you test it?

1. Add below records and see if the new customer and payment methods are created. try making a recurring payment with the payment method saved
```
merchant_id,merchant_connector_id,name,email,phone,phone_country_code,customer_id,subscription_id,payment_method,payment_method_type,nick_name,payment_instrument_id,card_number_masked,card_expiry_month,card_expiry_year,card_scheme,original_transaction_id,billing_address_zip,billing_address_state,billing_address_first_name,billing_address_last_name,billing_address_city,billing_address_country,billing_address_line1,billing_address_line2,raw_card_number,original_transaction_amount,original_transaction_currency
merchant_1720684811,mca_qqgmyxXRtsEjRLAUHBD4,noreal name,test@gmail.com,523698741,33,958709111eda444cdfdb2e8b7a3,900036959,card,credit,,4CBB27C05DE063AF598E,433026XXXXXX4675,03,2027,1,16153570198,94043,CA,NOREAL,NAME,Mountain View,US,1295 Charleston Rd,,,1000,USD
merchant_1720684811,mca_qqgmyxXRtsEjRLAUHBD4,john doe,test1@gmail.com,9159999999,1,dbc00f611eea04f024251790ff4,9000369592,card,credit,,F7A6049ADFE063AF598,378282XXXXX0005,03,2024,3,52127621321,700000,AK,NOREAL,NAME,hcm,US,hcm,,,,
```
```
curl --location 'http://localhost:8080/payment_methods/migrate-batch' \
--header 'api-key: test_admin' \
--header 'Content-Type: application/json' \
--form 'file=@"/Users/jagan.elavarasan/Downloads/sample.csv"'
```
<img width="1000" alt="image" src="https://github.com/user-attachments/assets/ba2e3ccc-ccf8-48d9-b148-931aea4cfe3f">

```
curl --location 'http://localhost:8080/customers/edca1edee0e411ec961c506b8dc695e7/payment_methods' \
--header 'Accept: application/json' \
--header 'x-feature: router-custom-be' \
--header 'api-key: dev_hRgXc8VdzglIywDYBmC7MBpw7dNGIAniCGRdqrF7cDdkWLoGi2iMfEI83pVGz2zz'
```
<img width="694" alt="image" src="https://github.com/user-attachments/assets/d96af05c-1b40-4fce-831d-eb220b4cb8d0">

```
curl --location 'http://localhost:8080/payments/pay_xAY8tQGtAijAhFX4a4pq/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: pk_dev_3a60113d6c90405d99a563e78961880b' \
--data '{
    "payment_method": "card",
    "client_secret": "pay_xAY8tQGtAijAhFX4a4pq_secret_AgGaWFPSW3inNKcxkHcg",
    "payment_token": "token_3VdfjykAKGS1ukxixQoM"
}'
```
<img width="627" alt="image" src="https://github.com/user-attachments/assets/45ad03c2-1947-491c-b841-b55bea3ad20c">

2. Try uploading big file > 1MB, it should throw appropriate error

```
curl --location 'http://localhost:8080/payment_methods/migrate-batch' \
--header 'api-key: test_admin' \
--header 'Content-Type: application/json' \
--form 'file=@"/Users/jagan.elavarasan/Downloads/sample.csv"'
```
<img width="595" alt="image" src="https://github.com/user-attachments/assets/47adf4a9-3c28-40a8-b715-e400238bed78">

3. Try uploading mulitple merchant at a time, it should not be allowed
<img width="555" alt="image" src="https://github.com/user-attachments/assets/a1af5dcb-168f-43b1-846c-af782333bcd7">

4. Should not allow merchant or merchant_connector_account if not created before
 <img width="567" alt="Screenshot 2024-07-12 at 15 06 50" src="https://github.com/user-attachments/assets/e3992488-b965-457d-810b-56c85db8dab6">
<img width="1012" alt="image" src="https://github.com/user-attachments/assets/95b8186a-06e8-4d1e-ad04-702e94703482">


## Checklist
